eks/core.py

```diff
@@ -461,198 +108,233 @@ def compute_initial_guesses(
+        s_finals: (K,) final `s` per keypoint (blockwise value broadcast to members).
+        ms: (K, T, D) smoothed state means.
+        Vs: (K, T, D, D) smoothed state covariances.
+
+    Notes:
+        • NLL is computed with EKF *filter*; outputs use EKF *smoother*.
+        • Loss for a block is the sum of member keypoints’ NLLs (via vmap).
+        • All jitted helpers close over optimizer/tol/bounds to avoid passing Python objects.
     """
-
-    n_keypoints = ys.shape[0]
-    s_finals = []
-    if blocks is None:
-        blocks = []
-    if len(blocks) == 0:
-        for n in range(n_keypoints):
-            blocks.append([n])
+    # -------------------- setup & time-varying R_t --------------------
+    K, T, obs_dim = ys.shape
+    if not blocks:
+        blocks = [[k] for k in range(K)]
     if verbose:
-        print(f'Correlated keypoint blocks: {blocks}')
-
-    @partial(jit)
-    def nll_loss_sequential_scan(s, cov_mats, cropped_ys, m0s, S0s, Cs, As, ensemble_vars):
-        s = jnp.exp(s)  # To ensure positivity
-        return smooth_min(
-            s, cov_mats, cropped_ys, m0s, S0s, Cs, As, ensemble_vars)
-
-    loss_function = nll_loss_sequential_scan
-
-    # Optimize smooth_param
+        print(f"Correlated keypoint blocks: {blocks}")
+
+    # Build time-varying R
+    Rs = build_R_from_vars(np.swapaxes(ensemble_vars, 0, 1))
+    Rs_j = jnp.asarray(Rs)
+
+    # Device arrays once
+    ys_j = jnp.asarray(ys)
+    m0s_j = jnp.asarray(m0s)
+    S0s_j = jnp.asarray(S0s)
+    As_j = jnp.asarray(As)
+    Qs_j = jnp.asarray(Qs)
+    Cs_j = jnp.asarray(Cs)
+
+    # Initial guesses per keypoint
+    s_guess_per_k = np.empty(K, dtype=float)
+    for k in range(K):
+        g = float(compute_initial_guesses(ensemble_vars[:, k, :]) or 2.0)
+        s_guess_per_k[k] = g if (np.isfinite(g) and g > 0.0) else 2.0
+
+    # -------------------- choose or optimize s --------------------
+    s_finals = np.empty(K, dtype=float)
     if smooth_param is not None:
-        if isinstance(smooth_param, float):
-            s_finals = [smooth_param]
-        elif isinstance(smooth_param, int):
-            s_finals = [float(smooth_param)]
+        if isinstance(smooth_param, (int, float)):
+            s_finals[:] = float(smooth_param)
         else:
-            s_finals = smooth_param
+            s_finals[:] = np.asarray(smooth_param, dtype=float)
     else:
-        guesses = []
-        cropped_ys = []
-        for k in range(n_keypoints):
-            current_guess = compute_initial_guesses(ensemble_vars[:, k, :])
-            guesses.append(current_guess)
-            if s_frames is None or len(s_frames) == 0:
-                cropped_ys.append(ys[k])
-            else:
-                cropped_ys.append(crop_frames(ys[k], s_frames))
-
-        cropped_ys = np.array(cropped_ys)  # Concatenation of this list along dimension 0
-
-        # Optimize negative log likelihood
+        optimizer = optax.adam(float(lr))
```

Comment:
let's put everything inside this `else` block inside it's own function; that way it'll be easier to follow the flow of this part of the code:
```
    if smooth_param is not None:
        s_finals = np.empty(K, dtype=float)
        if isinstance(smooth_param, (int, float)):
            s_finals[:] = float(smooth_param)
        else:
            s_finals[:] = np.asarray(smooth_param, dtype=float)
    else:
        s_finals = optimize_smooth_param(
            ...
        )
```